[🐛 Bug]: [java] Resource not found when using ExpectedConditions.visibilityOfElementLocated

### Description

When running latest Selenium 4.41 in an Java OSGi environment, `IllegalArgumentException` is thrown from method [`Read.resourceAsString`](https://github.com/SeleniumHQ/selenium/blob/trunk/java/src/org/openqa/selenium/io/Read.java#L56).

`Read.class.getResourceAsStream(resource)` can only load resources accessible from the class loader of this class. This currently limits it the selenium-api JAR/OSGi bundle.

One option would be to move the Read class into  selenium-remote-driver JAR/OSGi bundle where the resources (.js files) are located.

### Reproducible Code

```java
WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(10));
wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//body")));
```

### Debugging Logs

```java
org.openqa.selenium.remote.UnreachableBrowserException: 
Error communicating with the remote browser. It may have died.
Build info: version: '4.41.0', revision: '9fc754f'
System info: os.name: 'Windows Server 2022', os.arch: 'amd64', os.version: '10.0', java.version: '25.0.1'
Driver info: org.openqa.selenium.chrome.ChromeDriver
Command: [96fc25d881898671b536eeccf3c58ab9, isElementDisplayed [id]]
Capabilities {acceptInsecureCerts: false, browserName: chrome, browserVersion: 146.0.7680.80, chrome: {chromedriverVersion: 146.0.7680.165 (4b989da09e1...}, fedcm:accounts: true, goog:chromeOptions: {debuggerAddress: localhost:50001}, pageLoadStrategy: normal, platformName: any, proxy: Proxy(), se:cdp: ws://localhost:50001/devtoo..., se:cdpVersion: 146.0.7680.80, setWindowRect: false, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:extension:minPinLength: true, webauthn:extension:prf: true, webauthn:virtualAuthenticators: true}
Element: [[ChromeDriver: chrome on any (96fc25d881898671b536eeccf3c58ab9)] -> xpath: //body]
Session ID: 96fc25d881898671b536eeccf3c58ab9
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:640)
	at org.openqa.selenium.remote.RemoteWebElement.execute(RemoteWebElement.java:223)
	at org.openqa.selenium.remote.RemoteWebElement.isDisplayed(RemoteWebElement.java:282)
	at org.openqa.selenium.support.ui.ExpectedConditions.elementIfVisible(ExpectedConditions.java:365)
	at org.openqa.selenium.support.ui.ExpectedConditions$7.apply(ExpectedConditions.java:222)
	at org.openqa.selenium.support.ui.ExpectedConditions$7.apply(ExpectedConditions.java:215)
	at org.openqa.selenium.support.ui.FluentWait.until(FluentWait.java:209)
	at Test Code
Caused by: java.lang.IllegalArgumentException: Resource not found: /org/openqa/selenium/remote/isDisplayed.js
	at org.openqa.selenium.io.Read.resourceAsString(Read.java:59)
	at org.openqa.selenium.remote.codec.w3c.W3CHttpCommandCodec.lambda$executeAtom$2(W3CHttpCommandCodec.java:366)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1724)
	at org.openqa.selenium.remote.codec.w3c.W3CHttpCommandCodec.executeAtom(W3CHttpCommandCodec.java:362)
	at org.openqa.selenium.remote.codec.w3c.W3CHttpCommandCodec.amendParameters(W3CHttpCommandCodec.java:272)
	at org.openqa.selenium.remote.codec.AbstractHttpCommandCodec.encode(AbstractHttpCommandCodec.java:238)
	at org.openqa.selenium.remote.codec.AbstractHttpCommandCodec.encode(AbstractHttpCommandCodec.java:113)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:211)
	at org.openqa.selenium.remote.service.DriverCommandExecutor.invokeExecute(DriverCommandExecutor.java:216)
	at org.openqa.selenium.remote.service.DriverCommandExecutor.execute(DriverCommandExecutor.java:174)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:604)
	... 7 more
```

---
ℹ️ **Last known working version:** `4.39`

--- a/java/src/org/openqa/selenium/devtools/events/CdpEventTypes.java
+++ b/java/src/org/openqa/selenium/devtools/events/CdpEventTypes.java
@@ -66,7 +66,9 @@ public class CdpEventTypes {
   public static EventType<Void> domMutation(Consumer<@Nullable DomMutationEvent> handler) {
     Require.nonNull("Handler", handler);
 
-    String script = Read.resourceAsString("/org/openqa/selenium/devtools/mutation-listener.js");
+    String script =
+        Read.resourceAsString(
+            CdpEventTypes.class, "/org/openqa/selenium/devtools/mutation-listener.js");
 
     return new EventType<>() {
       @Override

--- a/java/src/org/openqa/selenium/io/Read.java
+++ b/java/src/org/openqa/selenium/io/Read.java
@@ -18,6 +18,7 @@
 package org.openqa.selenium.io;
 
 import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
 
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
@@ -53,8 +54,19 @@ public class Read {
     return new String(toByteArray(in), UTF_8);
   }
 
+  /**
+   * This method might not work in OSGI environment.
+   *
+   * @deprecated Use {@link #resourceAsString(Class, String)} instead
+   */
+  @Deprecated
   public static String resourceAsString(String resource) {
-    try (InputStream stream = Read.class.getResourceAsStream(resource)) {
+    return resourceAsString(Read.class, resource);
+  }
+
+  public static String resourceAsString(Class<?> resourceOwner, String resource) {
+    Class<?> clazz = requireNonNull(resourceOwner, "Class owning the resource");
+    try (InputStream stream = clazz.getResourceAsStream(resource)) {
       if (stream == null) {
         throw new IllegalArgumentException("Resource not found: " + resource);
       }

--- a/java/src/org/openqa/selenium/remote/RemoteScript.java
+++ b/java/src/org/openqa/selenium/remote/RemoteScript.java
@@ -84,7 +84,7 @@ class RemoteScript implements Script {
   @Override
   public long addDomMutationHandler(Consumer<DomMutation> consumer) {
     String scriptValue =
-        Read.resourceAsString("/org/openqa/selenium/remote/bidi-mutation-listener.js");
+        Read.resourceAsString(getClass(), "/org/openqa/selenium/remote/bidi-mutation-listener.js");
 
     this.script.addPreloadScript(scriptValue, List.of(new ChannelValue("channel_name")));
 

--- a/java/src/org/openqa/selenium/remote/codec/w3c/W3CHttpCommandCodec.java
+++ b/java/src/org/openqa/selenium/remote/codec/w3c/W3CHttpCommandCodec.java
@@ -363,7 +363,7 @@ public class W3CHttpCommandCodec extends AbstractHttpCommandCodec {
               atomFileName,
               (fileName) -> {
                 String rawFunction =
-                    resourceAsString("/org/openqa/selenium/remote/" + atomFileName);
+                    resourceAsString(getClass(), "/org/openqa/selenium/remote/" + atomFileName);
                 String atomName = fileName.replace(".js", "");
                 return String.format(
                     "/* %s */return (%s).apply(null, arguments);", atomName, rawFunction);

--- a/java/test/org/openqa/selenium/io/ReadTest.java
+++ b/java/test/org/openqa/selenium/io/ReadTest.java
@@ -42,7 +42,7 @@ class ReadTest {
 
   @Test
   void canReadResourceFromClasspath() {
-    String script = Read.resourceAsString("/org/openqa/selenium/remote/isDisplayed.js");
+    String script = Read.resourceAsString(Read.class, "/org/openqa/selenium/remote/isDisplayed.js");
     assertThat(script).isNotBlank().contains("function(){");
   }
 }

